### ISSUE TITLE
matrix parameters

### ISSUE THREAD

**Issue by mf59816** [Soostone/uri-bytestring#13]
Would you consider a pull request along these lines?:

```haskell
data URI
  { ...
  , uriPath :: Path
 ...
  }

newtype Path = Path [Segment]
data Segment = Segment ByteString [MatrixParam]
data MatrixParam = MatrixParam ByteString ByteString
```

(`Path` could also be a `type`.)

I'm not sure about the status of standardization of matrix parameters, so it may make sense to allow library users to block them in the parser, and maybe even in the types.  But it would make the library a lot more attractive for e.g. servant.

Any ideas or constraints?

thanks!  (-:

For discussion, see also: https://github.com/haskell-servant/servant/issues/139#issuecomment-114752961

**Comment by mf59816** [comment]
Re the question of compliance, http://tools.ietf.org/html/rfc3986#section-3.3 has this to say:

```
   Aside from dot-segments in hierarchical paths, a path segment is
   considered opaque by the generic syntax.  URI producing applications
   often use the reserved characters allowed in a segment to delimit
   scheme-specific or dereference-handler-specific subcomponents.  For
   example, the semicolon (";") and equals ("=") reserved characters are
   often used to delimit parameters and parameter values applicable to
   that segment.  The comma (",") reserved character is often used for
   similar purposes.  For example, one URI producer might use a segment
   such as "name;v=1.1" to indicate a reference to version 1.1 of
   "name", whereas another might use a segment such as "name,1.1" to
   indicate the same.  Parameter types may be defined by scheme-specific
   semantics, but in most cases the syntax of a parameter is specific to
   the implementation of the URI's dereferencing algorithm.
```

My interpretation of this is that it is legal for the path to contain matrix parameters, but it is not part of the standard.  So one option would be to stick with opaque path segments of type `ByteString` provide a function:

```haskell
matrixedSegment :: ByteString -> (ByteString, [MatrixParam])
```

Not sure what the implications for servant and its various users are?  My preference would be to instead expose the extra structure in the type as proposed and provide a functions:

```haskell
opaqueSegment :: Segment -> ByteString
opaquePath :: Path -> ByteString
```

**Comment by MichaelXavier** [comment]
The part that gives me the most pause about this is that the docs themselves indicate that there isn't really a standard way to parse this, as both semicolon and comma are valid separators for matrix params. How would you implement this parser?

One approach we could take is to keep the core types for URI on the more generic side. Taking this to the extreme would mean even removing query parameter pairs since the spec says they're opaque. I wouldn't advocate for this since almost every use case for uri-bytestring demands query parameter parsing, but in my view matrix parameters are a lot less common. We could however add some more detailed types and functions for segmenting paths and matrix params. So rather than parsing immediately into matrix params, have your `matrixedSegment` and maybe even a `pathSegments` function that would parse a path into stronger types. That would be completely opt-in, non-disruptive and could take a configuration parameter for configuring the matrix parser (similar to `URIParserOptions`) to use different delimiters.

I've never worked on an application that used matrix params and haven't heard about them until today, but I have a vague memory of seeing them in some URIs before, and the fact that Tim Berners-Lee wrote about it gives it some weight. I think the line for this library should be minimal disruption, minimal dependencies while providing a solid toolkit for users. So I'd be okay with adding support for matrix params in this way but wouldn't be okay with adding support for say some arbitrary protocol that something like Yesod uses. That isn't the goal of this library and could easily be implemented in a separate library, say `uri-bytestring-yesod`.

**Comment by mf59816** [comment]
Thanks for reasoning this out.  I don't want to admit this, but it makes a lot of sense.

I'll keep thinking a bit.  Hope it'll make sense to me once it's written down.

Paths in the RFC are actually segmented, so adding the `Path` type above would break the law much less than keeping the `Query` type already there.  I'm in favor of doing that, but perhaps there are other reasons not to (performance, backwards compatibility).

As for matrix params, what about this?:

```haskell
type URI = URI' Path
type URIMx = URI' PathMx
data URI' path = URI'
  { ...
  , uriPathMx :: path
  ...
  }

newtype Path = Path [ByteString]
newtype PathMx = PathMx [Segment]
data Segment = Segment ByteString [MatrixParam]
data MatrixParam = MatrixParam ByteString ByteString
  -- or use QueryParam instead of MatrixParam?  or (ByteString, ByteString)?
```

Then we could provide two parsers, one for `URI` and one for `URIMx`.

Does that change anything for the current users?

**Comment by mightybyte** [comment]
I agree with Michael that we should keep the core types as simple and generic as possible and provide extra functionality as optional layers on top.  One thing we could do to facilitate this is to keep the raw query string around for ease and efficiency of parsing by any of these optional layers.  We could do something like this:

    data URI = URI {
          uriScheme    :: Scheme
        , uriAuthority :: Maybe Authority
        , uriPath      :: ByteString
        , uriRawQuery  :: ByteString
        , uriQuery     :: Query
        , uriFragment  :: Maybe ByteString
        -- ^ URI fragment. Does not include the #
        }

Or instead of keeping both uriRawQuery and uriQuery we could get rid of uriQuery and provide a separate function to generate it.  If we did that, then the current Query type could be just one of many query string representations that we might make available to the user.  I don't have a strong opinion on whether the URI type should keep both or just uriRawQuery, but I do think it makes sense to have uriRawQuery so we don't have to concatenate Query back to a ByteString for the user to use an alternative query string parser.

**Comment by MichaelXavier** [comment]
I suppose if we dropped `uriQuery` from the record and replaced it with a function that would be more or less interface-compatible with what we have now (save for things like pattern matching on the URI record). It does have the nice factor of making additional query parsers conform to the same interface.

One problem I see is that this may open the door into people accidentally re-parsing the query every time  they use `uriQuery` and friends. For instance, if you had code that used `uriQuery u` in multiple places, would GHC typically be smart enough to recognize that as pure and memoize the result? Presumably that is a non issue for record fields since the value is precomputed.

I'm not a huge fan of parameterizing URI based on its features (i.e. having a path matrix). I don't think its a huge step backwards to use functions for extracting path segments vs using a type alias (aside from the aforementioned concerns about memoization).
